cranelift/egraph/src/lib.rs

```diff
@@ -256,6 +258,20 @@ pub trait Language: CtxEq<Self::Node, Self::Node> + CtxHash<Self::Node> {
     fn needs_dedup(&self, node: &Self::Node) -> bool;
 }
 
+/// A trait that allows the aegraph to compute a property of each
+/// node as it is created.
+pub trait Analysis {
+    type L: Language;
+    type Value: Debug + Clone + Default;
```

Comment:
> Would you mind double-checking for me that you need Analysis::Value to implement all three of Debug, Clone, and Default? I naively wouldn't have expected Default to be necessary since initial analysis results should come from Analysis::for_node.
>
> I assume both Default and Clone are because of the requirements for SecondaryMap.

Indeed, they both come from `SecondaryMap`'s constraints on its item type `V`; verified just now by attempting to remove them.

It turns out `Debug` was only needed for, well, my debugging (some `log`s I had in place); those are no longer present so I've removed that constraint. We can always put it back if we want to check in some persistent `log`s, I think.

> Then you could declare EGraph::nodes as a vector of (L::Node, A::Value), so that it's clear at the type level that if a node has been added, its analysis result exists too. And then you shouldn't need Default or Clone.

There is a nice appeal to this for sure! I actually had built the thing this way originally. I'm hesitant to go back to it chiefly because it makes the API for `for_node` more awkward: we need to provide it access to the analysis values for prior nodes (so it can depend on e.g. arguments' analysis values) and I don't like the idea of exposing internal data-structure details, like a tuple-of-node-and-value. Unfortunately we can't parameterize it on the "getter" because we can't provide a function trait-constrained type parameter to the trait method. And we can't just pass in the whole `EGraph` because the `EGraph` is parameterized on the particular `Analysis` and if I try to use `Self` there, I run into issues with unsized types (??). Basically, it becomes a type-system mess. So I'd kind of prefer to at least run with what we have for now and maybe explore more advanced solutions here later, if possible :-)

Reply:
I just had a go at doing this, and while I think it's feasible (with some `?Sized` constraints to allow unsized types if needed), the current implementation is tangled up enough in assumptions about analysis working this way that I don't want to try harder right now. So yes, let's come back to this later. :laughing: